File: Python3/539. Minimum Time Difference.py
class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        time = []
        for t in timePoints:
            h, m = t[:2], t[3:]
            time.append(int(h) * 60 + int(m))

        time.sort()
        ans = 780
        for i, t in enumerate(time[1:], 1):
            ans = min(ans, t - time[i - 1])

        return min(ans, 24 * 60 - time[-1] + time[0])

        preh, prem = time[0]
        #ans = min(ans, (curh - preh) * 60 + curm - perm)
        return ans

File: Java/539. Minimum Time Difference.java
class Solution {
    public int findMinDifference(List<String> timePoints) {
        boolean[] time = new boolean[24 * 60];
        for (String t: timePoints) {
            int min = Integer.parseInt(t.substring(0, 2)) * 60 + 
                      Integer.parseInt(t.substring(3));
            if (time[min]) {
                return 0;
            }
            time[min] = true;
        }

        int prev = -1, ans = 780, first = -1, last = -1;
        for (int i = 0; i < 24 * 60; i++) {
            if (time[i]) {
                if (prev != -1) {
                    ans = Math.min(ans, i - prev);
                }
                prev = i;

                // Update the index of the first valid time
                if (first == -1) {
                    first = i;
                }

                // Update the index of the last valid time every roll
                last = i;
            }
        }

        return Math.min(ans, 24 * 60 - last + first);
    }
}